### DISCUSSION TITLE
MapTransform for missing keys

### DISCUSSION THREAD

**Discussion by jansoe** [Project-MONAI/MONAI#1591] [Q&A] [answered]
The documentation for the abstract MapTransform  
https://github.com/Project-MONAI/MONAI/blob/8084394bce52aa1836ddb787ba6440f4362dc05b/monai/transforms/compose.py#L251-L257
states that MapTransform should "do nothing" for missing keys.
But all implemented MapTransform will fail if for example the `label` key is not available
https://github.com/Project-MONAI/MONAI/blob/8084394bce52aa1836ddb787ba6440f4362dc05b/monai/transforms/spatial/dictionary.py#L272-L274
Is there a rational for this?
Is there an elegant way to deal with missing labels e.g. in some mixed supervised and unsupervised training

**Comment by rijobro** [comment, marked as answer]
If you wanted to update the MONAI source code to do nothing for extra keys (it would require replacing all dictionary for loops), finding the intersection between the desired keys and the data's keys should work:

```python
for key in set(self.keys) & set(data.keys()):
     d[key] = self.rotator(d[key]) 
 return d 
```

If you wanted a solution that would work without modifying MONAI's code, perhaps you could create a dummy variable? For example, if `label` is missing, then create a small numpy image. You'd also want to record whether a dummy variable needed to be created (probably under another key e.g., `data["label_is_dummy"]=True`). You could do all that by creating your own callable transform. It sounds a little hacky to me, not sure if anyone else has a better idea.

**Reply by jansoe** [reply]
The main question would be, if such an update of the source code is desired or if developers rather consider the current behaviour as `fail fast`. 
In my opinion the non-existents of a key is a reasonable scenario in DL pipelines with semi-labelled data, and I don't have any idea how to setup such a pipeline without implementing the functionality in the transform pipeline.

In the end it would probably take the same effort to update the source code as implementing the dummy hack (which could actually increase the computational burden by a lot)
Maybe the easiest would be to create some decorator to add to all applicable classes?
```
@ignore_missing_keys
def __call__(self, data):
     ...
     for k in self.keys():
     ...
```

```
def ignore_missing_keys(original_call_method):      
    def replaced_call_method(self, data):  
        self.all_keys = self.keys
        self.keys = set(self.all_keys).intersection(data.keys())                   
        transformed = original_call_method(data)                
        self.keys = self.all_keys                      
        return transformed                                             
    return new_function
```
^^ not yet tested, but rough idea

**Reply by ericspod** [reply]
I would suggest that fail fast is the better design decision and will catch more errors than ignoring missing keys will solve. In the semi-labelled regime you would explicitly know that sometimes labels are missing, in this case providing a transform which fills in missing labels with dummy data would allow subsequent ones to operate as normal and would incur only a minor performance penalty if your dummy values are very small. This would make it explicit that you're accounting for a semi-labelled pipeline rather than leaving that to be implied behaviour.

**Reply by jansoe** [reply]
But than the documentation of the abstract MapTransform should definitely be updated to reflect that this is the desired behaviour:
(https://github.com/Project-MONAI/MONAI/blob/8084394bce52aa1836ddb787ba6440f4362dc05b/monai/transforms/compose.py#L251-L257)

**Reply by jansoe** [reply]
^^ @ericspod

**Reply by rijobro** [reply]
We've added it to our todo list: https://github.com/Project-MONAI/MONAI/issues/1662

**Comment by wyli** [comment]
this is a good point, perhaps we need to update the dict transforms to have an `allow_missing` boolean for each of the `keys`? @ericspod @Nic-Ma

**Comment by rijobro** [comment]
@jansoe This should be merged with https://github.com/Project-MONAI/MONAI/pull/1676. Give it a go and let us know if you have any problems!

The misleading documentation has also been updated: https://github.com/Project-MONAI/MONAI/blob/master/monai/transforms/transform.py#L176-L182.
